### PR TITLE
 cache disc executable to finish a MNIST model training

### PR DESCRIPTION
**Pull request by Yancey1989**
This PR adds a simple cache to train an MNIST with an MLP model as the following char illustration.
As you've noticed, this PR just makes Disc works as an LTC backend, but not work well, we still have some precision(**cannot converge to baseline**) and performance problems to solve.
But I prefer to merge this PR first as a baseline so that some new feature development and bugfix would be done in parallel, and of course, I will create some issues to trace these features and bugs.

image

### CODE REVIEW

File: torch_disc/torch_disc/csrc/disc_compiler/disc_compiler.cpp [hunk 1]
```diff
@@ -33,13 +34,12 @@ std::vector<torch::lazy::BackendDataPtr> Executable::Run(
     } else {
       // TODO(whc) should this check be made more general? it's written somewhat
       // oddly
-      CHECK(default_device_is_cuda ||
+      CHECK(!default_device_is_cuda ||
```
Comment (fortianyou) [inline torch_disc/torch_disc/csrc/disc_compiler/disc_compiler.cpp:37]: What does this mean?

File: torch_disc/torch_disc/csrc/disc_compiler/disc_compiler.cpp [hunk 2]
```diff
@@ -68,12 +68,18 @@ void EnhancementInputShape(
   }
 }
 
-std::shared_ptr<Executable> CompileToDiscExecutable(
+ExecutablePtr CompileToDiscExecutable(
     const std::shared_ptr<torch::jit::Graph>& graph,
     c10::ArrayRef<torch::lazy::BackendDataPtr> arguments) {
+  bool disable_disc = lazy_tensors::sys_util::GetEnvBool("DISABLE_DISC", false);
+  if (disable_disc) {
```
Comment (fortianyou) [inline torch_disc/torch_disc/csrc/disc_compiler/disc_compiler.cpp:75]: Is it used to testing TorchScript execution results?

File: torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp [hunk 3]
```diff
@@ -27,6 +27,7 @@ std::vector<Node*> FakeCluster(const std::shared_ptr<Graph>& graph) {
   for (auto node : graph->nodes()) {
     if (torch::blade::IsMlirMhloSupported(*node) &&
         node->kind() != prim::Constant) {
+      if (node->kind() == aten::addmm) continue;
```
Comment (fortianyou) [inline torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp:30]: Is the aten::addmm skipped because it running into accuracy error?

Comment (Yancey1989) [reply torch_disc/torch_disc/csrc/disc_compiler/passes/cluster.cpp:30]: no, `admm` caused unchanged loss, it is a bug, will fix it later https://github.com/alibaba/BladeDISC/issues/215

File: torch_disc/torch_disc/csrc/disc_compiler/disc_compiler.cpp (hunk 2 above)
Comment (Yancey1989) [reply torch_disc/torch_disc/csrc/disc_compiler/disc_compiler.cpp:75]: Yes.

File: torch_disc/torch_disc/csrc/disc_compiler/disc_compiler.cpp (hunk 1 above)
Comment (Yancey1989) [reply torch_disc/torch_disc/csrc/disc_compiler/disc_compiler.cpp:37]: just following the `TSBackendImpl` code, I'm not sure whether to delete them yet.

**Review by fortianyou — APPROVED** [review]
LGTM

**Pull request merged**